dm/worker/worker.go

```diff
@@ -165,119 +189,137 @@ func (w *Worker) Close() {
+	}
 
-	// we can remove this from SubTaskConfig later, because syncer will always read from relay
-	cfg.AutoFixGTID = w.cfg.AutoFixGTID
+	cfgStr, err := cfg.Toml()
+	if err != nil {
+		return 0, errors.Annotatef(err, "[worker] encode subtask %+v into toml format", cfg)
+	}
+	var opLogID int64
+	opLogID, err = w.operateSubTask(&pb.TaskMeta{
+		Op:   pb.TaskOp_Update,
+		Name: cfg.Name,
+		Task: append([]byte{}, cfgStr...),
+	})
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
 
-	// log config items, mydumper unit use it
-	cfg.LogLevel = w.cfg.LogLevel
-	cfg.LogFile = w.cfg.LogFile
+	return opLogID, nil
 }
 
 // StopSubTask stops a running sub task
-func (w *Worker) StopSubTask(name string) error {
-	if w.closed.Get() == closedTrue {
-		return errors.NotValidf("worker already closed")
-	}
-
+func (w *Worker) StopSubTask(name string) (int64, error) {
 	w.Lock()
 	defer w.Unlock()
-	st, ok := w.subTasks[name]
-	if !ok {
-		return errors.NotFoundf("sub task with name %s", name)
-	}
 
-	st.Close()
-	delete(w.subTasks, name)
-	return nil
-}
-
-// PauseSubTask pauses a running sub task
-func (w *Worker) PauseSubTask(name string) error {
 	if w.closed.Get() == closedTrue {
-		return errors.NotValidf("worker already closed")
+		return 0, errors.NotValidf("worker already closed")
 	}
 
-	st := w.findSubTask(name)
-	if st == nil {
-		return errors.NotFoundf("sub task with name %s", name)
+	var opLogID int64
+	opLogID, err := w.operateSubTask(&pb.TaskMeta{
+		Name: name,
+		Op:   pb.TaskOp_Stop,
+	})
+	if err != nil {
+		return 0, errors.Trace(err)
 	}
 
-	return st.Pause()
+	return opLogID, nil
 }
 
 // ResumeSubTask resumes a paused sub task
-func (w *Worker) ResumeSubTask(name string) error {
+func (w *Worker) ResumeSubTask(name string) (int64, error) {
+	w.Lock()
+	defer w.Unlock()
+
 	if w.closed.Get() == closedTrue {
-		return errors.NotValidf("worker already closed")
+		return 0, errors.NotValidf("worker already closed")
 	}
 
-	st := w.findSubTask(name)
-	if st == nil {
-		return errors.NotFoundf("sub task with name %s", name)
+	var opLogID int64
+	opLogID, err := w.operateSubTask(&pb.TaskMeta{
+		Name: name,
+		Op:   pb.TaskOp_Resume,
+	})
+	if err != nil {
+		return 0, errors.Trace(err)
 	}
 
-	return st.Resume()
+	return opLogID, nil
 }
 
-// UpdateSubTask update config for a sub task
-func (w *Worker) UpdateSubTask(cfg *config.SubTaskConfig) error {
+// PauseSubTask pauses a running sub task
+func (w *Worker) PauseSubTask(name string) (int64, error) {
+	w.Lock()
+	defer w.Unlock()
+
 	if w.closed.Get() == closedTrue {
-		return errors.NotValidf("worker already closed")
+		return 0, errors.NotValidf("worker already closed")
 	}
 
-	st := w.findSubTask(cfg.Name)
-	if st == nil {
-		return errors.NotFoundf("sub task with name %s", cfg.Name)
+	var opLogID int64
+	opLogID, err := w.operateSubTask(&pb.TaskMeta{
+		Name: name,
+		Op:   pb.TaskOp_Pause,
+	})
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+
+	return opLogID, nil
```

Comment:
extract similar code from `StopSubTask`/`PauseSubTask`/`ResumeSubTask` to a function has an arg named `op`?

Reply:
sure